File: backend/app/providers/sandbox.py
"""Supplier sandbox inventory — not real airlines.

Duffel test tokens issue ZZ / XX / YY (Duffel Airways). Drop those whenever
any real carrier exists. This is a supplier quirk, not a geography list.
"""

from __future__ import annotations

from typing import TYPE_CHECKING

if TYPE_CHECKING:
    from app.models import Offer

TEST_CARRIERS = frozenset({"ZZ", "XX", "YY"})


def is_sandbox_offer(offer: Offer) -> bool:
    if offer.live is False:
        return True
    note = offer.note or ""
    return offer.source == "duffel" and "live_mode=False" in note


def _carrier_codes(offer: Offer) -> set[str]:
    codes = {offer.carrier, offer.validating_airline}
    for segment in offer.segments:
        codes.add(segment.carrier)
        if segment.operating_carrier:
            codes.add(segment.operating_carrier)
    return {c.upper() for c in codes if c}


def is_live_fare(offer: Offer) -> bool:
    """True when the row is real published inventory, not mock or Duffel test."""
    if is_sandbox_offer(offer):
        return False
    if (offer.source or "").lower() == "mock":
        return False
    codes = _carrier_codes(offer)
    if codes and codes <= TEST_CARRIERS:
        return False
    if (offer.source or "").lower() == "duffel":
        return offer.live is True
    return True

File: backend/src/providers/sandbox.rs
use crate::models::Offer;
use std::collections::HashSet;

pub const TEST_CARRIERS: &[&str] = &["ZZ", "XX", "YY"];

pub fn is_test_carrier(code: &str) -> bool {
    TEST_CARRIERS.iter().any(|c| c.eq_ignore_ascii_case(code))
}

fn sandbox_note(note: &str) -> bool {
    note.to_ascii_lowercase().contains("live_mode=false")
}

pub fn is_sandbox_offer(offer: &Offer) -> bool {
    if offer.live == Some(false) {
        return true;
    }
    offer.source.eq_ignore_ascii_case("duffel") && sandbox_note(offer.note.as_deref().unwrap_or(""))
}

fn carrier_codes(offer: &Offer) -> HashSet<String> {
    let mut codes = HashSet::new();
    if !offer.carrier.is_empty() {
        codes.insert(offer.carrier.to_uppercase());
    }
    if let Some(v) = &offer.validating_airline {
        if !v.is_empty() {
            codes.insert(v.to_uppercase());
        }
    }
    for segment in &offer.segments {
        if !segment.carrier.is_empty() {
            codes.insert(segment.carrier.to_uppercase());
        }
        if let Some(op) = &segment.operating_carrier {
            if !op.is_empty() {
                codes.insert(op.to_uppercase());
            }
        }
    }
    codes
}

pub fn is_live_fare(offer: &Offer) -> bool {
    if is_sandbox_offer(offer) {
        return false;
    }
    if offer.source.eq_ignore_ascii_case("mock") {
        return false;
    }
    let codes = carrier_codes(offer);
    if !codes.is_empty() && codes.iter().all(|c| is_test_carrier(c)) {
        return false;
    }
    if offer.source.eq_ignore_ascii_case("duffel") {
        return offer.live == Some(true);
    }
    false
}
